modules/llrt_crypto/src/subtle/decrypt.rs

```diff
@@ -0,0 +1,77 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
+use aes_gcm::{aead::Aead, KeyInit, Nonce};
+use ctr::{cipher::StreamCipher, Ctr128BE, Ctr32BE, Ctr64BE};
+use llrt_utils::result::ResultExt;
+use rquickjs::{Ctx, Exception, Result};
+use rsa::{pkcs1::DecodeRsaPrivateKey, Oaep, RsaPrivateKey};
+use sha2::Sha256;
+
+use crate::subtle::{Aes256Gcm, Algorithm};
+
+type Aes256CbcDec = cbc::Decryptor<aes::Aes256>;
+
+pub fn decrypt(
+    ctx: &Ctx<'_>,
+    algorithm: &Algorithm,
+    key: Vec<u8>,
+    data: Vec<u8>,
+) -> Result<Vec<u8>> {
+    match algorithm {
+        Algorithm::AesGcm(iv) => {
+            let cipher = Aes256Gcm::new_from_slice(&key).or_throw(ctx)?;
+            let nonce = Nonce::from_slice(iv);
+
+            match cipher.decrypt(nonce, data.as_ref()) {
+                Ok(result) => Ok(result),
+                Err(_) => Err(Exception::throw_message(ctx, "Decryption failed"))?,
+            }
+        },
+        Algorithm::AesCbc(iv) => {
+            match Aes256CbcDec::new(key.as_slice().into(), iv.as_slice().into())
+                .decrypt_padded_vec_mut::<Pkcs7>(&data)
+            {
+                Ok(result) => Ok(result),
+                Err(_) => Err(Exception::throw_message(ctx, "Decryption failed")),
+            }
+        },
+        Algorithm::AesCtr(counter, length) => match length {
+            32 => decrypt_aes_ctr_gen::<Ctr32BE<aes::Aes256>>(ctx, &key, counter, &data),
+            64 => decrypt_aes_ctr_gen::<Ctr64BE<aes::Aes256>>(ctx, &key, counter, &data),
+            128 => decrypt_aes_ctr_gen::<Ctr128BE<aes::Aes256>>(ctx, &key, counter, &data),
+            _ => Err(Exception::throw_message(
+                ctx,
+                "invalid counter length. Currently supported 32/64/128 bits",
+            )),
+        },
+        Algorithm::RsaOaep(label) => {
+            let private_key = RsaPrivateKey::from_pkcs1_der(&key).or_throw(ctx)?;
+            let padding = match label {
+                Some(buf) => {
+                    Oaep::new_with_label::<Sha256, String>(String::from_utf8(buf.to_vec())?)
+                },
+                None => Oaep::new::<Sha256>(),
+            };
```

Comment:
Since "?" cannot be used in map, it looks like you will need to nest `map(...).unwrap_or_else(...)`.

```rust
let padding = label
    .as_ref()
    .map(|buf| {
        String::from_utf8(buf.to_vec())
            .map(|s| Oaep::new_with_label::<Sha256, String>(s))
            .unwrap_or_else(|_| Oaep::new::<Sha256>())
    })
    .unwrap_or_else(|| Oaep::new::<Sha256>());
```
Is it worth doing?

Reply:
I think we can even do this. We don't need UTF8 validations. In fact, we probably don't need UTF8 validations for any of the subtle crypto string conversions.
```rust
let padding = label.map_or(
    Oaep::new::<Sha256>(),
    |buf| Oaep::new_with_label::<Sha256>(String::from_utf8_lossy(buf).into_owned())
);
```